app/models/study_file.rb

```diff
@@ -1331,6 +1333,86 @@ def set_cluster_group_ranges
+
+            # collect the previous data fragment ids to choose the appropriate job for new or existing clusterings
+            prev_ids = []
+            self.ann_data_file_info.data_fragments_was.each do |data_frag|
+              prev_ids << data_frag&.[]("_id")
+            end
+
+          self.update(parse_status: 'parsing')
+
+          # going through the obsm key names ensures parse jobs are only being called on clusterings
+          self.ann_data_file_info.obsm_key_names.each do |fragment|
+              Rails.logger.info "launching AnnData #{fragment} cluster extraction"
+              matcher = { data_type: :cluster, obsm_key_name: fragment }
+              cluster_data_fragment = self.ann_data_file_info.find_fragment(**matcher)
+              
+              # set reparse to true only if the clustering already exists and is being updated
+              to_reparse = false
+              if prev_ids.include? cluster_data_fragment&.[]("_id")
+                to_reparse = true
+              end
+
+              # if new clustering need to extract first
+              if to_reparse == false 
+                params_object = AnnDataIngestParameters.new(
+                  anndata_file: self.gs_url, extract: %w[cluster], obsm_keys: [fragment]
+                )    
+                job = IngestJob.new(
+                  study:, study_file: self, user: study.user, action: :ingest_anndata, params_object:
+                )
+                job.delay.push_remote_and_launch_ingest
+
+              else
+                # primarily copied from launch_anndata_subparse_jobs in ingest_job 
+                name = cluster_data_fragment&.[](:name) || fragment # fallback if we can't find data_fragment
+                cluster_gs_url = fragment_file_gs_url(study.bucket_id, 'cluster', self.id, fragment)
+                domain_ranges = self.ann_data_file_info.get_cluster_domain_ranges(name).to_json
+                cluster_params = AnnDataIngestParameters.new(
+                  ingest_cluster: true, name:, cluster_file: cluster_gs_url, domain_ranges:, ingest_anndata: false,
+                  extract: nil, obsm_keys: nil
+                )
+
+                job = IngestJob.new(study:, study_file: self, user: study.user, action: :ingest_cluster, reparse: to_reparse, params_object: cluster_params)
+                job.delay.push_remote_and_launch_ingest
+              end
+            end
+
+            self.update(parse_status: 'parsed')
+
+          end
+        end
+      end
+    end
+
+  # Copied from ann_data_ingest_parameters.rb
+  # generate a GS URL to a derived fragment that was extracted from the parent AnnData file
+  # File name structure is: <input_filetype>_frag.<file_type>.<file_type_detail>.tsv
+  #   file_type = cluster|metadata|matrix
+  #   file_type_detail [optional] = cluster name (for cluster files), raw|processed (for matrix files)
+  def fragment_file_gs_url(bucket_id, fragment_type, h5ad_file_id, file_type_detail = "")
```

Comment:
This looks like a copy of `AnnDataIngestParameters#fragment_file_gs_url` - we should reference that method rather than copying it here.  That being said, if we do end up going the above route, then we wouldn't need this here.